src/libraries/System.Private.CoreLib/src/System/IO/FileStream.cs

```diff
@@ -594,304 +428,130 @@ public override void SetLength(long value)
 
-            // We now have space in the buffer. Store the byte.
-            GetBuffer()[_writePos++] = value;
+        ~FileStream()
+        {
+            // Preserved for compatibility since FileStream has defined a
+            // finalizer in past releases and derived classes may depend
+            // on Dispose(false) call.
+            Dispose(false);
         }
 
-        /// <summary>
-        /// Validates that we're ready to write to the stream,
-        /// including flushing a read buffer if necessary.
-        /// </summary>
-        private void PrepareForWriting()
+        protected override void Dispose(bool disposing)
         {
-            if (_fileHandle.IsClosed)
-                throw Error.GetFileNotOpen();
-
-            // Make sure we're good to write.  We only need to do this if there's nothing already
-            // in our write buffer, since if there is something in the buffer, we've already done
-            // this checking and flushing.
-            if (_writePos == 0)
+            if (_impl != null) // possible in finalizer
             {
-                if (!CanWrite) throw Error.GetWriteNotSupported();
-                FlushReadBuffer();
-                Debug.Assert(_bufferLength > 0, "_bufferSize > 0");
+                _impl.DisposeInternal(disposing);
             }
         }
 
-        ~FileStream()
+        public override ValueTask DisposeAsync()
         {
-            // Preserved for compatibility since FileStream has defined a
-            // finalizer in past releases and derived classes may depend
-            // on Dispose(false) call.
-            Dispose(false);
+            if (OperatingSystem.IsWindows() && GetType() != typeof(FileStream))
+            {
+                return base.DisposeAsync();
+            }
+
+            return _impl.DisposeAsync();
         }
 
+        public override Task CopyToAsync(Stream destination, int bufferSize, CancellationToken cancellationToken) => _impl.CopyToAsync(destination, bufferSize, cancellationToken);
+
         public override IAsyncResult BeginRead(byte[] buffer, int offset, int count, AsyncCallback? callback, object? state)
         {
             ValidateBufferArguments(buffer, offset, count);
-            if (IsClosed) throw new ObjectDisposedException(SR.ObjectDisposed_FileClosed);
-            if (!CanRead) throw new NotSupportedException(SR.NotSupported_UnreadableStream);
+            if (_impl.IsClosed) throw new ObjectDisposedException(SR.ObjectDisposed_FileClosed);
+            if (!_impl.CanRead) throw new NotSupportedException(SR.NotSupported_UnreadableStream);
 
-            if (!IsAsync)
+            if (!_impl.IsAsync)
                 return base.BeginRead(buffer, offset, count, callback, state);
             else
-                return TaskToApm.Begin(ReadAsyncTask(buffer, offset, count, CancellationToken.None), callback, state);
+                return _impl.BeginRead(buffer, offset, count, callback, state);
```

Comment:
Why do we need this if/else?  The ctor is already creating a specialized instance based on whether it's being constructed as async or not.  A key reason for the strategy pattern is to then avoid the need for this kind of if/else throughout the codebase, as the _impl already knows whether it's sync or async and how to handle a sync or async operation accordingly.